migrations/20210204190556-add-slug-to-projects.js

```diff
@@ -0,0 +1,97 @@
+const async = require('async');
+
+function slugify(title) {
+  return title
+    .replace(/[^a-zA-Z\d\s_()-./\\]/g, '')
+    .replace(/(\s|_|\(|\)|\/|\\|\.)+/g, '-')
+    .toLowerCase();
+}
+
+async function createSlug(id, title, collection) {
+  const slug = slugify(title);
+  let realSlug;
+  let count = 0;
+  let postfix = 0;
+  await async.doWhilst(
+    cb => {
+      realSlug = postfix === 0 ? slug : `${slug}-${postfix + 1}`;
+      collection
+        .countDocuments({
+          slug: realSlug,
+          _id: { $ne: id },
+        })
+        .then(_count => {
+          count = _count;
+          cb();
+        })
+        .catch(err => {
+          cb(err);
+        });
+      postfix += 1;
+    },
+    testCb => {
+      testCb(null, count > 0);
+    },
+  );
+  return realSlug;
+}
+
+module.exports = {
+  async up(db, client) {
+    // add slug to dacs
+    const dacs = await db
+      .collection('dacs')
+      .find({})
+      .toArray();
+    await async.eachLimit(dacs, 1, (dac, cb) => {
+      const { title } = dac;
+      createSlug(dac._id, title, db.collection('dacs')).then(slug => {
+        db.collection('dacs')
+          .updateOne({ _id: dac._id }, { $set: { slug } })
+          .then(() => {
+            cb();
+          });
+      });
+    });
+    // add slug to campaigns
+    const campaigns = await db
+      .collection('campaigns')
+      .find({})
+      .toArray();
+    await async.eachLimit(campaigns, 1, (campaign, cb) => {
+      const { title } = campaign;
+      createSlug(campaign._id, title, db.collection('campaigns')).then(slug => {
+        db.collection('campaigns')
+          .updateOne({ _id: campaign._id }, { $set: { slug } })
+          .then(() => {
+            cb();
+          });
+      });
+    });
+
+    // add slug to milestones
+    const milestones = await db
+      .collection('milestones')
+      .find({})
+      .toArray();
```

Comment:
I was concerned about large number of milestones cause an issue and run it on my machine hopefully it was OK. Personally use mongo cursor for such situation which has option to set parallelism, just as a suggestion for next time :)